src/components/application_manager/src/commands/hmi/get_urls.cc

```diff
@@ -48,48 +48,136 @@ void GetUrls::Run() {
   using namespace strings;
   using namespace hmi_apis;
 
-  SmartObject& object = *message_;
-  object[params][message_type] = MessageType::kResponse;
   if (!ApplicationManagerImpl::instance()->GetPolicyHandler().PolicyEnabled()) {
-    object[params][hmi_response::code] = Common_Result::DATA_NOT_AVAILABLE;
-    ApplicationManagerImpl::instance()->ManageHMICommand(message_);
+    SendResponseToHMI(Common_Result::DATA_NOT_AVAILABLE);
     return;
   }
 
+  SmartObject& object = *message_;
+  const std::string service_to_check =
+      object[msg_params][hmi_request::service].asString();
+
   policy::EndpointUrls endpoints;
   ApplicationManagerImpl::instance()->GetPolicyHandler().GetServiceUrls(
-      object[msg_params][hmi_request::service].asString(), endpoints);
+      service_to_check, endpoints);
+
   if (endpoints.empty()) {
-    object[params][hmi_response::code] = Common_Result::DATA_NOT_AVAILABLE;
+    LOG4CXX_ERROR(logger_, "No URLs for service " << service_to_check);
+    SendResponseToHMI(Common_Result::DATA_NOT_AVAILABLE);
+    return;
+  }
+
+#ifdef EXTENDED_POLICY
+  const std::string policy_service = "7";
+
+  if (policy_service == service_to_check) {
+    ProcessPolicyServiceURLs(endpoints);
+    return;
+  }
+#endif
+  ProcessServiceURLs(endpoints);
+}
+
+#ifdef EXTENDED_POLICY
+void GetUrls::ProcessPolicyServiceURLs(const policy::EndpointUrls& endpoints) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  using namespace smart_objects;
+  using namespace application_manager;
+  using namespace strings;
+  using namespace hmi_apis;
+
+  const uint32_t app_id_to_send_to =
+      ApplicationManagerImpl::instance()->
+      GetPolicyHandler().GetAppIdForSending();
+
+  if (!app_id_to_send_to) {
+    LOG4CXX_ERROR(logger_,
+                  "There are no available applications for processing.");
     ApplicationManagerImpl::instance()->ManageHMICommand(message_);
     return;
   }
 
+  ApplicationSharedPtr app =
+      ApplicationManagerImpl::instance()->application(app_id_to_send_to);
+
+  if (!app.valid()) {
+    LOG4CXX_WARN(logger_,
+                 "There is no registered application with "
+                 "connection key '" << app_id_to_send_to << "'");
+    SendResponseToHMI(Common_Result::DATA_NOT_AVAILABLE);
+    return;
+  }
+
+  SmartObject& object = *message_;
   object[msg_params].erase(hmi_request::service);
   object[msg_params][hmi_response::urls] = SmartObject(SmartType_Array);
 
   SmartObject& urls = object[msg_params][hmi_response::urls];
```

Comment:
do you really need this variable here. I can see that first usage is only after 30 lines ?